### PR TITLE
Allow multiple nonlinear systems for the same FE problem/mesh

### LINKED ISSUE #8780: Enable MOOSE to partially solve a problem
### Description of the enhancement or error report
Suppose a problem has two primal variables u and v. We can ask FEProblem to solve u with a fixed v. Of course we can solve v with a fixed u. It's noted that both u and v are in the same FEProblem, so v can be coupled in kernels of u and vise versa. I suggest to create a function FEProblem::solve (std::vector<NonlinearVariableName>). PJFNK is still desired for those partial solves.

### Rationale for the enhancement or information for reproducing the error
This of cause can be implemented with MultiApp/Transfer, but it owns advantages MultiApp cannot provide:
1. No data transfer and extra storage for the transferred data;
2. No need to duplicate mesh;
3. Materials can be shared by both variables;
4. Enable much cleaner code if complicated iterative scheme is required.

This will clean up implementations of an acceleration scheme in Rattlesnake dramatically and pave a way of more advanced schemes which has been proved effective.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

New feature which will make Rattlesnake transport calculations more memory and CPU time efficiently and a much better user interface (no need to create two input files with multiapps).

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #8780

### CODE REVIEW

**Comment by hugary1995** [comment]
I just read through the issue you referenced. It was a fun one to read actually. Though it seems that a few people are against of having multiple systems (mainly due to difficulty in maintenance?). Hopefully the picture has changed in the past few years.

**Comment by hugary1995** [comment]
Also, I'm onboard with this design. I'm happy to provide my use cases when needed.

**Comment by friedmud** [comment]
I assume that one of the people that was against it in the past that is for it is me ;-)

When we discussed this last time things were in a very different state.  We had a lot of work to do on MultiApps to get it where we needed it.  It's now settled down quite a bit.  In addition, many new use-cases for this new capability have emerged (including CFD).

As for this PR: It's incredibly well done!  Very well thought out - and well implemented.  Great job @lindsayad !

**Comment by loganharbour** [comment]
There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not

**Comment by lindsayad** [comment]
> There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not

This is an example of where it's appropriate
- we reinit all assembly objects with a given element
- we now want to resize something, say stateful material properties. We can pull any assembly's quadrature rule and it should have the same number of quadrature points for resizing. In such an example, pulling at index 0 is as good as any

**Comment by loganharbour** [comment]
> > There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not
> 
> This is an example of where it's appropriate
> 
>     * we reinit all assembly objects with a given element
> 
>     * we now want to resize something, say stateful material properties. We can pull any assembly's quadrature rule and it should have the same number of quadrature points for resizing. In such an example, pulling at index 0 is as good as any

Is there a reason why we wouldn't want to just use the actual system from the primary variable?

File: framework/src/base/Assembly.C [hunk 1]
```diff
@@ -1110,11 +1043,10 @@ Assembly::computeSinglePointMapAD(const Elem * elem,
         libmesh_assert(elem_nodes[i]);
         const Node & node = *elem_nodes[i];
         libMesh::VectorValue<DualReal> elem_point = node;
-        unsigned dimension = 0;
-        if (computingJacobian())
-          for (const auto & disp_num : _displacements)
+        if (do_derivatives)
+          for (const auto & [disp_num, direction] : _disp_numbers_and_directions)
```
Comment (loganharbour) [inline framework/src/base/Assembly.C:1047]: hnnnnng it's beautiful

**Comment by loganharbour** [comment]
> > > There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not
> > 
> > 
> > This is an example of where it's appropriate
> > ```
> > * we reinit all assembly objects with a given element
> > 
> > * we now want to resize something, say stateful material properties. We can pull any assembly's quadrature rule and it should have the same number of quadrature points for resizing. In such an example, pulling at index 0 is as good as any
> > ```
> 
> Is there a reason why we wouldn't want to just use the actual system from the primary variable?
> 
> Edit: oh wait... this is aux

Okay you can ignore this - I'm on board.

Almost through, nothing so far. FEProblemBase is a beast... will finish up this afternoon

**Review by loganharbour — APPROVED** [review]

**Comment by lindsayad** [comment]
Thank you for the review!!!

**Pull request merged**